### PR TITLE
Regularization normaleqn dev

### PR DESCRIPTION
**Pull request by nychiang**
Always update dual regularizations first if normal equation is used. 
Note that by default linear system `xdycyd`, we update primal first.

The current implementation has some numerical issue when randomized regularizations is used.
That is, the BICG iterative method and some other functions still use the unified regularization in computation. 
Will have a follow-up PR to fix this issue.

### CODE REVIEW

File: src/LinAlg/hiopVectorPar.cpp [hunk 1]
```diff
@@ -124,6 +125,22 @@ void hiopVectorPar::setToConstant(double c)
...
+
+#ifdef NDEBUG
+  re.seed(std::chrono::system_clock::now().time_since_epoch().count());
+#endif
+
+  for(index_type i=0; i<n_local_; ++i) {
+    data_[i] = unif(re);
+  }
+}
+
```
Comment (nychiang) [inline src/LinAlg/hiopVectorPar.cpp:142]: Any good way to implement this function in hiopVectorRAJAPar?
@cnpetra @ashermancinelli

Comment (ashermancinelli) [reply src/LinAlg/hiopVectorPar.cpp:142]: cuRand is shipped in cuda toolkit, but we might have to add some linker flags.

Comment (ashermancinelli) [reply src/LinAlg/hiopVectorPar.cpp:142]: [We could do almost exactly what this Thrust example is doing with RAJA](https://docs.nvidia.com/cuda/curand/device-api-overview.html#thrust-and-curand-example).

Comment (nychiang) [reply src/LinAlg/hiopVectorPar.cpp:142]: @ashermancinelli I know we have cuRAND and rocRAND, but the problem now is that in this case I cannot see a portable solution to both cuda and hip. It seems to me that we need to use `#ifdef HIOP_USE_CUDA` with `cuRAND`, and `#ifdef HIOP_USE_HIP` with `rocRAND`.

Comment (ashermancinelli) [reply src/LinAlg/hiopVectorPar.cpp:142]: I see now. Grepping through raja's docs doesn't find anything for random, so I think ugly macros are the only way to go.

Comment (cnpetra) [reply src/LinAlg/hiopVectorPar.cpp]: not much to add -- it appears we need to rely on roc and cuda api (with `ifdef`s). We will add an abstraction layer for device math functions call later on.

File: src/LinAlg/hiopVectorPar.hpp [hunk 2]
```diff
@@ -70,6 +70,7 @@ class hiopVectorPar : public hiopVector
 
   virtual void setToZero();
   virtual void setToConstant( double c );
+  virtual void set_to_random_constant(double minv, double maxv);
```
Comment (cnpetra) [inline src/LinAlg/hiopVectorPar.hpp:73]: "constant" is a bit confusing maybe? name need to reflect that the uniform distribution is sampled. Did you mean `set_to_random_uniform` ?

File: src/LinAlg/hiopVectorRajaPar.hpp [hunk 3]
```diff
@@ -77,6 +77,7 @@ class hiopVectorRajaPar : public hiopVector
 
   virtual void setToZero();
   virtual void setToConstant( double c );
+  virtual void set_to_random_constant(double minv, double maxv) {assert(false&&"not implemented yet!");}
```
Comment (cnpetra) [inline src/LinAlg/hiopVectorRajaPar.hpp:80]: we need this implementation rather sooner than later. A quick follow-up PR would be a good idea

**Review by cnpetra — APPROVED** [review]
will merge as-is ... name of vector's random function and the device implementation can be done via a follow-up PR

**Pull request merged**